### PR TITLE
Fix hybrid metadata

### PR DESCRIPTION
**Pull request by Scienfitz**
Performs a small refactoring in recommenders to enable hybrid space metadata marking which was previously ignored

### CODE REVIEW

File: baybe/recommenders/base.py [hunk 1]
```diff
@@ -55,40 +55,20 @@ class Recommender(ABC, RecommenderProtocol):
...
-
-        if searchspace.type == SearchSpaceType.DISCRETE:
-            # Select, recommend, and mark discrete candidates in one atomic step
-            return self._select_candidates_and_recommend(
-                searchspace.discrete,
-                batch_size,
-            )
-
-        if searchspace.type == SearchSpaceType.CONTINUOUS:
+        """See base class."""
```
Comment (AVHopp) [inline baybe/recommenders/base.py:65]: Please change this back. Using the exception results in the method inheriting the actual docstring of the parent class. This solution replaces it by `See base class` which is definitely inferior.

File: baybe/recommenders/base.py [hunk 2]
```diff
@@ -193,16 +173,18 @@ def _recommend_hybrid(
...
 
-    def _select_candidates_and_recommend(
+    def _recommend_with_discrete_parts(
         self,
-        subspace_discrete: SubspaceDiscrete,
+        searchspace: SearchSpace,
         batch_size: int,
     ) -> pd.DataFrame:
-        """Get candidates in a discrete search space and generate recommendations.
+        """Obtain recommendations in search spaces with discrete parts.
```
Comment (AVHopp) [inline baybe/recommenders/base.py:181]: "with a discrete part", search spaces do not have mutliple discrete parts.

File: baybe/recommenders/base.py [hunk 3]
```diff
@@ -193,16 +173,18 @@ def _recommend_hybrid(
...
+
+        Convenience helper which sequentially performs the following tasks: get discrete
```
Comment (AVHopp) [inline baybe/recommenders/base.py:183]: I personally would prefer proper sentences here, but since this is a private function, I'd also be happy with this.

File: baybe/recommenders/base.py [hunk 4]
```diff
@@ -193,16 +173,18 @@ def _recommend_hybrid(
...
+        candidates, generate recommendations, update metadata.
 
         Args:
-            subspace_discrete: The discrete subspace from which to generate
-                recommendations.
+            searchspace: The searchspace from which to generate recommendations.
```
Comment (AVHopp) [inline baybe/recommenders/base.py:187]: search space

File: baybe/recommenders/base.py [hunk 5]
```diff
@@ -212,15 +194,20 @@ def _select_candidates_and_recommend(
...
                 recommendation than requested.
         """
+        is_hybrid_space = searchspace.type is SearchSpaceType.HYBRID
+
         # Get discrete candidates
-        _, candidates_comp = subspace_discrete.get_candidates(
-            allow_repeated_recommendations=self.allow_repeated_recommendations,
-            allow_recommending_already_measured=self.allow_recommending_already_measured,
+        _, candidates_comp = searchspace.discrete.get_candidates(
+            allow_repeated_recommendations=True
```
Comment (AVHopp) [inline baybe/recommenders/base.py:201]: According to myGPT, this statement can be simplified as `allow_repeated_recommendations = is_hybrid_space or self.allow_repeated_recommendations`, and I also think that this is indeed equivalent.
However, if you prefer the current version (which is more readable) then feel free to not change this in this function.

File: baybe/recommenders/base.py [hunk 6]
```diff
@@ -212,15 +194,20 @@ def _select_candidates_and_recommend(
...
+            else self.allow_repeated_recommendations,
+            allow_recommending_already_measured=True
+            if is_hybrid_space
+            else self.allow_recommending_already_measured,
         )
 
         # Check if enough candidates are left
-        # TODO [15917]: This check is not perfectly correct.
-        if len(candidates_comp) < batch_size:
+        if (not is_hybrid_space) and (len(candidates_comp) < batch_size):
```
Comment (AVHopp) [inline baybe/recommenders/base.py:210]: Why did you remove the TODO? Isn't the check basically still the same?

**Review by AVHopp — CHANGES_REQUESTED** [review]
Some minor things should be changed, but overall this looks good.

File: baybe/recommenders/base.py (hunk 6 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:210]: I wasnt sure whats not correct here hence assumed this todo is old, do you know?

One thing that comes to mind is: if len(candidates) < batch size one could still (trivially) recommend the remaining points. But that is not really conform with the function returning exactly `batch size` points as per current definition and assumption

File: baybe/recommenders/base.py (hunk 5 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:201]: like it and will change

File: baybe/recommenders/base.py (hunk 4 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:187]: changed

File: baybe/recommenders/base.py (hunk 3 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:183]: the sentence seems proper to me so unless you point out an alternative I will keep it

File: baybe/recommenders/base.py (hunk 2 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:181]: depends on whether you count parts of parts also as parts :D

changed

File: baybe/recommenders/base.py (hunk 1 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:65]: changed

this is till a very bad way of fixing the issue, no one (eg external people) will understand or remember unless repeatedly doing it this way

File: baybe/recommenders/base.py (hunk 6 above)
Comment (AVHopp) [reply baybe/recommenders/base.py:209]: I also think that this was the original reason for this TODO, but imo, the check and the corresponding error message is correct. So we can remove it here or add a small comment in the code about "There might in general still be some points but not enough". Do as you like and then resolve :)

**Review by AVHopp — APPROVED** [review]

File: baybe/recommenders/base.py (hunk 1 above)
Comment (AVHopp) [reply baybe/recommenders/base.py]: Nope, we actually mention this in our guidelines: 
image

File: baybe/recommenders/base.py (hunk 6 above)
Comment (Scienfitz) [reply baybe/recommenders/base.py:209]: yes i know, clearly it hasn't helped me in remembering it anyway :) which is of course on me

My point: it will happen again to others (or me :D )

Comment (AdrianSosic) [reply baybe/recommenders/base.py:209]: Hey guys, sorry, don't want to disturb, just saw this on my phone (really need to disable GH push messages now :D). But the TODO was not about having too few points. It's about the fact that technically checking the len of the "candidates" is not enough. Example: get_candidates gives me 2 points that are left in the searchspace, and I request 5. If "allow_repeated_recommendations" is `True`, I can still return 5 recommendations without problem by choosing them "mit zurücklegen"

Comment (Scienfitz) [reply baybe/recommenders/base.py:209]: so if the additional three recommendations are copies of the remaining two and recommended solely for the purpose of having a total of 5 - whats the point of that?  just to satisfy the docstring? Seems like a no-usecase for me

so imo this check is complete and the TODO obsolete

Keep in mind that this will only happen anyway if someone measures all combinations... which is not a use case of baybe

The error we throw is even suable in simulations module to detect that the simulation cannot cotinue

but to unblock us Ill restore the old TODO and merge

Comment (AdrianSosic) [reply baybe/recommenders/base.py:210]: I think what you suggested (just restore and continue) sounds reasonable. And while not strictly necessary in most use cases, I wouldn't say it's irrelevant. Especially in high noise situations, it's a valid question to ask which option to measure further. By saying this is irrelevant, you basically give the finger to the entire bandits community 😄

Comment (Scienfitz) [reply baybe/recommenders/base.py:210]: not so fast, I never said that the idea of duplicate recommendations per se is irrelevant, I agree that in the situation of noisy measurements an algorithm recommending duplicates due to noise-awareness would make a lot of sense

However, that is not the situation this TODO refers to, and that is not why this code would recommend duplicates. It would only do so to fill in the amount so it recommends to the batch size -> this has nothing to do with the above -> and imo doesnt make sense -> hence I think the todo is obsolete (your idea is better captured in a work item and not this todo)

btw the TODO references a number which I thought could be a work item but I wasnt able to find it

Comment (Scienfitz) [reply baybe/recommenders/base.py:210]: resolving because it was `reintroduced`
